soroban-env-host/src/host.rs

```diff
@@ -2077,7 +2077,16 @@ impl VmCallerEnv for Host {
         let scv = self.visit_obj(b, |hv: &ScBytes| {
             self.metered_from_xdr::<ScVal>(hv.as_slice())
         })?;
-        self.to_host_val(&scv)
+        if Val::can_represent_scval(&scv) {
+            self.to_host_val(&scv)
+        } else {
+            Err(self.err(
+                ScErrorType::Value,
+                ScErrorCode::InvalidAction,
+                "Deserialized ScVal type cannot be represented as Val",
+                &[(scv.discriminant() as i32).into()],
+            ))
+        }
```

Comment:
Should we instead change the error type inside the `check_val_representable_scval`, instead of copying its code? 
Of the 4 places it's being called, 3 seems to correspond to the same error type. The only one that would be an internal error is `from_host_val`, and we can `map_err` for it.

Reply:
I don't think so. I think the other sites are not accessible to user controlled input so it's properly an internal error for a non-val-representatable ScVal to show up. The embedder should have prevented it.

Reply:
The check as it is implemented here is non-exhaustive; instead it should happen recursively for every produced `ScVal`. A test where we try to deserialize e.g. a vector with `LedgerKeyContractInstance` element would help.

So probably we should do what Jay suggests: update `check_val_representable_scval` to return `Value, InvalidAction` error and convert it to `InternalError` where appropriate.

Reply:
Oh hah! I think the lack of recursion in is_val_representable is itself a bug. I'll fix it along with this.

I don't want to start mapping errors _out_ of `InternalError` since it risks hiding or mis-classifying other internal errors. Since 3 sites it's called are self-consistency checks -- not points of exposure to user input -- I really do think they should remain `InternalError`.

(For consistency of reporting I'm actually going to change this to `UnexpectedType`, to match the the low-resolution `ConversionError` projected errors we get from `TryFromVal` on a non-representable case going through conversion)